### PR TITLE
backport: merge bitcoin#22278, #24103, #24235, #24177, #24299, #24917, #22564, #25349, #25571, #17487, #26999, #27011 (blockstorage backports: part 2)

### PR DESCRIPTION
**Pull request by kwvg**
## Additional Information

* Dependent on https://github.com/dashpay/dash/pull/6097
* Dependency for https://github.com/dashpay/dash/pull/6067
* Test failures in `linux64_multiprocess-build` ([build](https://gitlab.com/dashpay/dash/-/jobs/7550924662)) and `linux64_tsan-test` ([build](https://gitlab.com/dashpay/dash/-/jobs/7550924666)) do not stem from this PR but are pre-existing failures in `develop` ([build](https://gitlab.com/dashpay/dash/-/jobs/7550859495), [build](https://gitlab.com/dashpay/dash/-/jobs/7550859499)). A fix for the build failures has been opened as a separate PR.

## Breaking Changes

None observed.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas **(note: N/A)**
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation **(note: N/A)**
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

**Comment by PastaPastaPasta** [comment]
Rebasing on develop

**Comment by PastaPastaPasta** [comment]
```
 -:  ---------- >  1:  5394d63e18 feat: improve merge-check action to leave a comment and label to PRs
 -:  ---------- >  2:  35c76705d1 feat: implement basic Github Actions based CI, which reuses underlying logic from GitLab CI
 -:  ---------- >  3:  d1addb27aa fix: change fallback download path to be an s3 link which includes a few packages
 -:  ---------- >  4:  ed8ffa7fb4 feat: have cppcheck linter respect CACHE_DIR env variable
 -:  ---------- >  5:  4dbdecdd1e refactor: rename builder-image -> build-image and builder as image name to dashcore-ci-runner
 -:  ---------- >  6:  acf1315270 Merge bitcoin/bitcoin#25091: test: Remove extended lint (cppcheck)
 -:  ---------- >  7:  2237b4266b refactor: use `=` with ENV in dockerfiles
 -:  ---------- >  8:  1b139e4837 feat: automatically run guix-build on all tags pushed
 -:  ---------- >  9:  101a31555f refactor: simplify caching setup, add a restore key to actually cache besides 1 run
 -:  ---------- > 10:  580bbe6d1c feat: improve guix building; run always, save artifacts
 -:  ---------- > 11:  770651aa15 set hosts in guix-check
 -:  ---------- > 12:  92f82f6fc6 fix: group the ENV in quotations
 -:  ---------- > 13:  aafded67d9 fix: compilation error due to rebase error between bitcoin#22937 and ipc/process
 -:  ---------- > 14:  540f6871d3 fix: lock `::cs_main` before accessing `ChainstateManager::m_best_header`
 -:  ---------- > 15:  2777d30caa fix(ci): only run some GH actions on PR approval, not on every comment
 -:  ---------- > 16:  c2bd834f3b Merge bitcoin/bitcoin#21169: fuzz: Add RPC interface fuzzing. Increase fuzzing coverage from 65% to 70%.
 -:  ---------- > 17:  4dd36f603a Merge bitcoin/bitcoin#21798: fuzz: Create a block template in tx_pool targets
 -:  ---------- > 18:  813993d44a Merge bitcoin/bitcoin#21892: fuzz: Avoid excessively large min fee rate in tx_pool
 -:  ---------- > 19:  2398283ff6 Merge bitcoin/bitcoin#21922: fuzz: Avoid timeout in EncodeBase58
 -:  ---------- > 20:  3bfefde53b Merge bitcoin/bitcoin#21931: ci: Bump cirrus fuzz CPUs to avoid timeout
 -:  ---------- > 21:  82a6aa5907 Merge bitcoin/bitcoin#21810: fuzz: Various RPC fuzzer follow-ups
 -:  ---------- > 22:  d89847f43e Merge bitcoin/bitcoin#22004: fuzz: Speed up transaction fuzz target
 -:  ---------- > 23:  06ea87040b Merge bitcoin/bitcoin#20773: refactor: split CWallet::Create
 -:  ---------- > 24:  4125485401 fix: follow-up bitcoin#20773 - for collateral lock/unlock coins
 -:  ---------- > 25:  9ad26f1664 fix: follow-up bitcoin#20773 - coinjoin loader can be nullptr too
 1:  c0a5f3c08c = 26:  3fddb84ff4 test: simplify `pow_test`'s `get_next_work block` index construction
 2:  f161a14ac9 = 27:  66490224a5 merge bitcoin#22278: Add LIFETIMEBOUND to CScript where needed
 3:  4b275bd5e3 = 28:  5d82ed7476 merge bitcoin#24103: Replace RecursiveMutex m_cs_chainstate with Mutex, and rename it
 4:  a231d04e3f = 29:  0d457178b4 merge bitcoin#24235: use stronger EXCLUSIVE_LOCKS_REQUIRED()
 5:  ec720b865f = 30:  7a0f22e20a merge bitcoin#24177: add missing thread safety lock assertions
 6:  4be170c323 = 31:  12d6d03e09 merge bitcoin#24299: UnloadBlockIndex and ChainstateManager::Reset thread safety cleanups
 7:  93f06ee075 = 32:  4ce6aa5bcb merge bitcoin#24917: Make BlockManager::LoadBlockIndex private
 8:  94d529e58d = 33:  ea01f62b81 partial bitcoin#22564: Move mutable globals cleared in ::UnloadBlockIndex to BlockManager
 9:  591c837a0c = 34:  464edf8c5d merge bitcoin#25349: CBlockIndex/CDiskBlockIndex improvements for safety, consistent behavior
10:  31afcc41cb = 35:  88f407439a merge bitcoin#25571: Make mapBlocksUnknownParent local, and rename it
11:  7e798d9808 = 36:  5d852c3e23 merge bitcoin#17487: allow write to disk without cache drop
12:  e86dce77a2 = 37:  b119bd4e0d merge bitcoin#26999: A few follow-ups to bitcoin#17487
```

**Review by PastaPastaPasta — COMMENTED** [review]
LGTM b119bd4e0d6fc145496c7f574d4b9d8fab6b34c9, but maybe should extract the first commit? I don't understand its purpose here

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: src/init.cpp [hunk 1]
```diff
@@ -1568,8 +1569,8 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     if (!ignores_incoming_txs) node.fee_estimator = std::make_unique<CBlockPolicyEstimator>();
 
     assert(!node.mempool);
-    int check_ratio = std::min<int>(std::max<int>(args.GetArg("-checkmempool", chainparams.DefaultConsistencyChecks() ? 1 : 0), 0), 1000000);
-    node.mempool = std::make_unique<CTxMemPool>(node.fee_estimator.get(), check_ratio);
+    const int mempool_check_ratio = std::clamp<int>(args.GetArg("-checkmempool", chainparams.DefaultConsistencyChecks() ? 1 : 0), 0, 1000000);
+    node.mempool = std::make_unique<CTxMemPool>(node.fee_estimator.get(), mempool_check_ratio);
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/init.cpp]: Resolved in latest push!

File: src/test/pow_tests.cpp [hunk 2]
```diff
@@ -11,112 +11,40 @@
...
+    { 1408728124, 0x1b104be1U }, { 1408728332, 0x1b10e09eU }, { 1408728479, 0x1b11a33cU },
+    { 1408728495, 0x1b121cf3U }, { 1408728608, 0x1b11951eU }, { 1408728744, 0x1b11abacU },
+    { 1408728756, 0x1b118d9cU }, { 1408728950, 0x1b1123f9U }, { 1408729116, 0x1b1141bfU },
+    { 1408729179, 0x1b110764U }, { 1408729305, 0x1b107556U }, { 1408729474, 0x1b104297U },
+    { 1408729576, 0x1b1063d0U }, { 1408729587, 0x1b10e878U }, { 1408729647, 0x1b0dfaffU },
+    { 1408729678, 0x1b0c9ab8U }, { 1408730179, 0x1b0c03d6U }, { 1408730862, 0x1b0dd168U },
+    { 1408730914, 0x1b10b864U }, { 1408731242, 0x1b0fed89U }, { 1408731256, 0x1b113ff1U },
+    { 1408732229, 0x1b10460bU }, { 1408732257, 0x1b13b83fU }, { 1408732489, 0x1b1418d4U }
+};
+
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/test/pow_tests.cpp]: Removed in latest push!

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK b65863783f0cb0ef9ad01284ee0c89adc2fac99b

**Review by PastaPastaPasta — APPROVED** [review]
utACK b65863783f0cb0ef9ad01284ee0c89adc2fac99b

**Review by knst — COMMENTED** [review]
post-utACK

27011 has missing changes for taproot

**Pull request merged**